# Used car analysis

We'll work with a dataset of used car ads scraped from eBay. This dataset was supplied by Dataquest.

In [2]:
import pandas as pd
import numpy as np
import re

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [3]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Pro observation: **we got some columns** 

In [6]:
autos_columns = autos.columns
autos_columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.rename({
    'yearOfRegistration':'registration_year',
    'monthOfRegistration':'registration_month',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'dateCrawled': 'date_crawled',
    'offerType': 'offer_type',
    'vehicleType': 'vehicle_type',
    'powerPS': 'power_ps',
    'fuelType': 'fuel_type',
    'nrOfPictures': 'num_of_pictures',
    'postalCode': 'postal_code',
    'lastSeen': 'last_seen',
}, inplace=True, axis='columns')

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Renamed all of the columns to snakecase to make them easier to navigate and also to suit column name best practices

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### num_of_pictures seems to be zero for all of its columns. postal_code is stored as a number even though it should be recognised as a string. registration_year seems to have a range from 1000 to 9999. I don't know anybody in the year 9999.... yet

### price should be stored as a number but it's stored as a string. odometer is also stored as a string. seller and offer_type have almost all of its values as the same

In [10]:
autos['price'].value_counts().head(10)

$0        1421
$500       781
$1,500     734
$2,500     643
$1,200     639
$1,000     639
$600       531
$800       498
$3,500     498
$2,000     460
Name: price, dtype: int64

In [11]:
autos['price'] = autos['price'].str.replace("$", "").str.replace(",", "").astype(int)

In [12]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [13]:
autos['odometer'] = autos['odometer'].str.replace(",", "").str.replace("km", "").astype(int)

In [14]:
autos.rename({'odometer': 'odometer_km'}, inplace=True, axis='columns')

In [15]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Let's take a closer look at the odometer_km column

In [16]:
autos['odometer_km'].unique().shape

(13,)

In [17]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [18]:
autos['odometer_km'].value_counts(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

### Doesn't seem to be too many outliers. Let's explore the price column

In [19]:
autos['price'].unique().shape

(2357,)

In [20]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

### Use format(x, 'f') in order to suppress scientific notation

In [21]:
autos['price'].describe().apply(lambda x: format(x, 'f'))

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price, dtype: object

### It seems like there's a LOAD of outliers sitting at the top of the price column

In [22]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [23]:
autos['price'].value_counts().sort_index(ascending=False).tail(20)

35       1
30       7
29       1
25       5
20       4
18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price, dtype: int64

### there are a ton of really REALLY cheap cars. These should be considered outliers and stripped

Let's remove cars sold for free and anything over 350k. Really, I should be stripping a lot more out of the bottom

In [24]:
autos = autos[autos['price'].between(1,350000)]

In [25]:
autos.shape

(48565, 20)

### We're left with 48.5k cars

In [26]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [27]:
autos['date_crawled'].isnull().value_counts(normalize=True, dropna=False)

False    1.0
Name: date_crawled, dtype: float64

### No null values. Pages were crawled between 5th March and 7th April

In [28]:
autos['date_crawled'].describe()

count                   48565
unique                  46882
top       2016-03-23 18:39:34
freq                        3
Name: date_crawled, dtype: object

In [29]:
autos['ad_created'].describe()

count                   48565
unique                     76
top       2016-04-03 00:00:00
freq                     1887
Name: ad_created, dtype: object

In [30]:
autos['ad_created'].str[:10].value_counts(dropna=False).sort_index(ascending=False).tail(20)

2016-02-08    1
2016-02-07    1
2016-02-05    2
2016-02-02    2
2016-02-01    1
2016-01-29    1
2016-01-27    3
2016-01-22    1
2016-01-16    1
2016-01-14    1
2016-01-13    1
2016-01-10    2
2016-01-07    1
2016-01-03    1
2015-12-30    1
2015-12-05    1
2015-11-10    1
2015-09-09    1
2015-08-10    1
2015-06-11    1
Name: ad_created, dtype: int64

In [31]:
autos['ad_created'].str[:10].value_counts(dropna=False).head(20)

2016-04-03    1887
2016-03-20    1843
2016-03-21    1825
2016-04-04    1790
2016-03-12    1785
2016-03-14    1709
2016-04-02    1707
2016-03-28    1699
2016-03-07    1687
2016-03-29    1653
2016-03-15    1652
2016-03-19    1636
2016-04-01    1636
2016-03-30    1627
2016-03-08    1618
2016-03-09    1610
2016-03-11    1598
2016-03-22    1593
2016-03-26    1567
2016-03-23    1557
Name: ad_created, dtype: int64

### Majority of ads were created between March and April. 

In [32]:
autos['last_seen'].str[:10].describe(include='all')

count          48565
unique            34
top       2016-04-06
freq           10772
Name: last_seen, dtype: object

In [33]:
autos['last_seen'].str[:10].value_counts(dropna=False).tail(10)

2016-03-14    612
2016-03-11    601
2016-03-10    518
2016-03-09    466
2016-03-13    432
2016-03-08    360
2016-03-18    357
2016-03-07    262
2016-03-06    210
2016-03-05     52
Name: last_seen, dtype: int64

In [34]:
autos['last_seen'].str[:10].value_counts(dropna=False).head(10)

2016-04-06    10772
2016-04-07     6408
2016-04-05     6059
2016-03-17     1364
2016-04-03     1224
2016-04-02     1210
2016-03-30     1203
2016-04-04     1189
2016-03-12     1155
2016-03-31     1155
Name: last_seen, dtype: int64

### Majority of ads last seen in April and March

In [35]:
autos['registration_year'].describe(include='all')

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Apparently some were registered in the year 1000 and 9999.... Let's strip out those mothafuckas

In [36]:
autos['registration_year'].value_counts().sort_index(ascending=False).head(20)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
2014     663
2013     803
2012    1310
2011    1623
2010    1589
Name: registration_year, dtype: int64

Can strip out everything after 2018. There's only a couple from 2019

In [37]:
autos['registration_year'].value_counts().sort_index(ascending=False).tail(70)

2004    2703
2003    2699
2002    2486
2001    2636
2000    3156
1999    2897
1998    2363
1997    1951
1996    1373
1995    1227
1994     629
1993     425
1992     370
1991     339
1990     347
1989     174
1988     135
1987      72
1986      72
1985      95
1984      51
1983      51
1982      41
1981      28
1980      85
1979      34
1978      44
1977      22
1976      21
1975      18
        ... 
1964      12
1963       8
1962       4
1961       6
1960      23
1959       6
1958       4
1957       2
1956       4
1955       2
1954       2
1953       1
1952       1
1951       2
1950       3
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       5
1800       2
1111       1
1001       1
1000       1
Name: registration_year, Length: 70, dtype: int64

### Need to work on outlier recognition. I'm not sure how to use box plots yet so let's just strip out everything before 1960

In [38]:
autos = autos[autos['registration_year'].between(1960,2018)]

In [39]:
autos.shape

(48496, 20)

In [40]:
autos['registration_year'].describe(include='all')

count    48496.000000
mean      2003.517795
std          7.329811
min       1960.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2018.000000
Name: registration_year, dtype: float64

In [41]:
autos['registration_year'].value_counts().sort_index()

1960      23
1961       6
1962       4
1963       8
1964      12
1965      17
1966      22
1967      26
1968      26
1969      19
1970      38
1971      26
1972      33
1973      23
1974      24
1975      18
1976      21
1977      22
1978      44
1979      34
1980      85
1981      28
1982      41
1983      51
1984      51
1985      95
1986      72
1987      72
1988     135
1989     174
1990     347
1991     339
1992     370
1993     425
1994     629
1995    1227
1996    1373
1997    1951
1998    2363
1999    2897
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
Name: registration_year, dtype: int64

### All of the years remaining in the dataset are believable. The majority of the cars are in the 90s-2010s

In [42]:
top_brands = autos['brand'].value_counts().head(20).index

In [43]:
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini'],
      dtype='object')

In [44]:
autos['brand'].describe(include='all')

count          48496
unique            40
top       volkswagen
freq           10329
Name: brand, dtype: object

### Brand contains the brand of car in the dataset. I've decided to explore the data for the top 20 listed in the dataset and find the mean price for those brands

In [45]:
brand_mean_price = {}
for brand in top_brands:
    brand_mean_price[brand] = autos[autos['brand'] == brand]['price'].mean()

In [46]:
# .items() returns a tuple of dictionary values. By using the lambda function one can sort the items by value

sorted(brand_mean_price.items(), key=lambda x: x[1], reverse=True)

[('sonstige_autos', 12228.629711751662),
 ('mini', 10541.566985645934),
 ('audi', 9212.9306621881),
 ('mercedes_benz', 8465.090047393365),
 ('bmw', 8260.768253366205),
 ('skoda', 6360.545571245186),
 ('hyundai', 5371.792960662526),
 ('volkswagen', 5329.804434117533),
 ('toyota', 5148.0032733224225),
 ('volvo', 4866.993166287016),
 ('nissan', 4669.3859649122805),
 ('seat', 4324.8725490196075),
 ('mazda', 4059.059539918809),
 ('citroen', 3739.6359649122805),
 ('ford', 3693.2141587677725),
 ('smart', 3518.102305475504),
 ('peugeot', 3065.611888111888),
 ('opel', 2940.9134724857686),
 ('fiat', 2793.8700475435817),
 ('renault', 2431.9840791738384)]

### Sonstige_autos are mad expensive, fam. Renault comes in at the cheapest

In [47]:
autos[autos['brand']=='sonstige_autos']['price'].value_counts().sort_index(ascending=False).head(10)

345000    1
194000    1
114400    1
109999    1
105000    1
80000     1
79999     1
79500     1
72600     1
62000     1
Name: price, dtype: int64

### The 345000 car could be considered an outlier which is why the sonstige_autos cost is coming out on top

In [48]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [49]:
autos['odometer_km'].value_counts()

150000    31407
125000     5056
100000     2109
90000      1731
80000      1414
70000      1213
60000      1152
50000      1011
40000       812
5000        810
30000       780
20000       758
10000       243
Name: odometer_km, dtype: int64

In [50]:
brand_mean_mileage = {}
for brand in top_brands:
    brand_mean_mileage[brand] = autos[autos['brand'] == brand]['odometer_km'].mean()

In [51]:
brand_mean_mileage

{'volkswagen': 128956.33652822151,
 'bmw': 132707.1875592642,
 'opel': 129482.92220113851,
 'mercedes_benz': 130975.87246876347,
 'audi': 129492.56238003839,
 'ford': 124478.67298578199,
 'renault': 128115.31841652324,
 'peugeot': 127356.64335664336,
 'fiat': 117567.35340729002,
 'seat': 121895.4248366013,
 'skoda': 111020.53915275994,
 'nissan': 118711.20107962213,
 'mazda': 124871.44790257105,
 'smart': 100511.52737752162,
 'citroen': 120190.05847953216,
 'toyota': 116219.31260229132,
 'hyundai': 106718.4265010352,
 'sonstige_autos': 92117.51662971175,
 'volvo': 138337.1298405467,
 'mini': 88899.52153110047}

In [52]:
brand_mileage = pd.Series(brand_mean_mileage)
brand_mileage

volkswagen        128956.336528
bmw               132707.187559
opel              129482.922201
mercedes_benz     130975.872469
audi              129492.562380
ford              124478.672986
renault           128115.318417
peugeot           127356.643357
fiat              117567.353407
seat              121895.424837
skoda             111020.539153
nissan            118711.201080
mazda             124871.447903
smart             100511.527378
citroen           120190.058480
toyota            116219.312602
hyundai           106718.426501
sonstige_autos     92117.516630
volvo             138337.129841
mini               88899.521531
dtype: float64

In [53]:
df = pd.DataFrame(brand_mileage, columns=['mean_mileage'])
brand_price = pd.Series(brand_mean_price)

In [54]:
df

mean_mileage
volkswagen      128956.336528
bmw             132707.187559
opel            129482.922201
mercedes_benz   130975.872469
audi            129492.562380
ford            124478.672986
renault         128115.318417
peugeot         127356.643357
fiat            117567.353407
seat            121895.424837
skoda           111020.539153
nissan          118711.201080
mazda           124871.447903
smart           100511.527378
citroen         120190.058480
toyota          116219.312602
hyundai         106718.426501
sonstige_autos   92117.516630
volvo           138337.129841
mini             88899.521531

In [55]:
df['mean_price'] = brand_price
brand_price

volkswagen         5329.804434
bmw                8260.768253
opel               2940.913472
mercedes_benz      8465.090047
audi               9212.930662
ford               3693.214159
renault            2431.984079
peugeot            3065.611888
fiat               2793.870048
seat               4324.872549
skoda              6360.545571
nissan             4669.385965
mazda              4059.059540
smart              3518.102305
citroen            3739.635965
toyota             5148.003273
hyundai            5371.792961
sonstige_autos    12228.629712
volvo              4866.993166
mini              10541.566986
dtype: float64

In [56]:
df

mean_mileage    mean_price
volkswagen      128956.336528   5329.804434
bmw             132707.187559   8260.768253
opel            129482.922201   2940.913472
mercedes_benz   130975.872469   8465.090047
audi            129492.562380   9212.930662
ford            124478.672986   3693.214159
renault         128115.318417   2431.984079
peugeot         127356.643357   3065.611888
fiat            117567.353407   2793.870048
seat            121895.424837   4324.872549
skoda           111020.539153   6360.545571
nissan          118711.201080   4669.385965
mazda           124871.447903   4059.059540
smart           100511.527378   3518.102305
citroen         120190.058480   3739.635965
toyota          116219.312602   5148.003273
hyundai         106718.426501   5371.792961
sonstige_autos   92117.516630  12228.629712
volvo           138337.129841   4866.993166
mini             88899.521531  10541.566986

In [57]:
df['mileage_to_price'] = df['mean_mileage'] / df['mean_price']

In [58]:
df

mean_mileage    mean_price  mileage_to_price
volkswagen      128956.336528   5329.804434         24.195322
bmw             132707.187559   8260.768253         16.064751
opel            129482.922201   2940.913472         44.028131
mercedes_benz   130975.872469   8465.090047         15.472472
audi            129492.562380   9212.930662         14.055523
ford            124478.672986   3693.214159         33.704699
renault         128115.318417   2431.984079         52.679341
peugeot         127356.643357   3065.611888         41.543629
fiat            117567.353407   2793.870048         42.080466
seat            121895.424837   4324.872549         28.184744
skoda           111020.539153   6360.545571         17.454562
nissan          118711.201080   4669.385965         25.423300
mazda           124871.447903   4059.059540         30.763640
smart           100511.527378   3518.102305         28.569814
citroen         120190.058480   3739.635965         32.139508
toyota          116219.312602   5148.003273         22.575610
hyundai         106718.426501   5371.792961         19.866444
sonstige_autos   92117.516630  12228.629712          7.532939
volvo           138337.129841   4866.993166         28.423531
mini             88899.521531  10541.566986          8.433236

In [59]:
df.sort_values(by='mileage_to_price', ascending=False)

mean_mileage    mean_price  mileage_to_price
renault         128115.318417   2431.984079         52.679341
opel            129482.922201   2940.913472         44.028131
fiat            117567.353407   2793.870048         42.080466
peugeot         127356.643357   3065.611888         41.543629
ford            124478.672986   3693.214159         33.704699
citroen         120190.058480   3739.635965         32.139508
mazda           124871.447903   4059.059540         30.763640
smart           100511.527378   3518.102305         28.569814
volvo           138337.129841   4866.993166         28.423531
seat            121895.424837   4324.872549         28.184744
nissan          118711.201080   4669.385965         25.423300
volkswagen      128956.336528   5329.804434         24.195322
toyota          116219.312602   5148.003273         22.575610
hyundai         106718.426501   5371.792961         19.866444
skoda           111020.539153   6360.545571         17.454562
bmw             132707.187559   8260.768253         16.064751
mercedes_benz   130975.872469   8465.090047         15.472472
audi            129492.562380   9212.930662         14.055523
mini             88899.521531  10541.566986          8.433236
sonstige_autos   92117.516630  12228.629712          7.532939

### Renault has the best price:mileage

In [60]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [61]:
autos.groupby(['brand', 'model']) \
    .size() \
    .reset_index() \
    .rename({0: 'count'}, axis='columns') \
    .sort_values('count', ascending=False) \
    .head(10)

brand     model  count
268     volkswagen      golf   3898
23             bmw       3er   2686
274     volkswagen      polo   1688
183           opel     corsa   1679
180           opel     astra   1410
272     volkswagen    passat   1383
13            audi        a4   1256
142  mercedes_benz  c_klasse   1161
24             bmw       5er   1150
145  mercedes_benz  e_klasse    977

### Volkswagen Golf is the most common combination, closely followed by VW Polo